documentation build: Use Python package at install location in cmake build and improve Python doc builder

luator opened the pull request:
## Description

Point to the install location of the Python package instead of the source location in Sphinx' `conf.py` and when calling `sphinx-apidoc`. This way modules that are generated at build time (e.g. using pybind11) will be considered.


This fixes two related but distinct issues:

- `sphinx-apidoc` needs to use the install location, otherwise pybind11 modules will not be listed in the generated documentation.
- `sphinx-build` needs to use the install location (done through the change in `conf.py`) as otherwise the pybind11 modules will not be importable at build-time, resulting in even pure Python packages which import them internally to cause an error and thus to be excluded from the generated documentation.

WIP:
Before merging, the following points need to be considered:
- [ ] The change in `conf.py` will break the Python builder of the documentation.  Not sure how to best handle this.  Maybe by simply adding an `if/else`?  I think in the long term, we should actually get rid of this duplicate setup (maybe by having a standalone documentation builder which can be called both from Python and CMake or even independent of the build?).
- [ ] Just checking for the existence of `PYTHON_INSTALL_DIR` internally is not the best solution, maybe better use an argument to the macro.
- [ ] We need to ensure that the documentation is built _after_ all Python modules are built/installed.  I think right now, it may fail on a clean rebuild.  Is there a way to achieve this without manually specifying dependencies?


## How I Tested

By building the documentation of the robot_fingers package (using cmake).



## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [x] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [ ] All new functions/classes are documented and existing documentation is updated according to changes.
- [x] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

luator commented:
What do you think about the open questions?

MaximilienNaveau commented:
good question I will have a look at it a bit later...

vincentberenz commented:
I confess being quite lost on all this.

MaximilienNaveau commented:
One can create CMake commands POST_BUILD:
`add_custom_command(TARGET mylibrary POST_BUILD
    COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_FILE:mylibrary> ${targetfile}
)`
This should help creating the doc after everything is build without specific dependencies.
I was looking for this as well when I did this setup but I failed to...

luator commented:
I started some effort to make the Python documentation builder usable as a standalone application and also some fixes and refactoring (still work in progress, though)

luator commented:
I finished my work on the Python documentation builder for now (I still have some small improvements in mind but I think they should better go to a separate PR).

Relevant changes I made:
- Fixed a bug in detecting C++ files (it never detected them previously)
- Add `:caption:` to the toctrees for a more organised side menu
- Use `sphinx_apidoc --separate`, so each module gets its own page
- Optional argument to specify path to Python package.  This allows pointing to the package in the `install/` directory so pybind11 modules are included.  If not set, it will search in the source directory by default as before.
- Also allow `docs/` (instead of `doc/`) for the general documentation.
- Some refactoring here and there
- I only discovered `mpi_doc_build` after I already added a `__main__` section in the module itself.  I think there is some merit in having everything in one file, so I changed `mpi_doc_build` to simply be an alias for calling the module with `python -m`.  This change the interface of mpi_doc_build`, though, not sure if that is a problem.

I tested with the following packages and the outcome looked good to me in all cases:
- robot_fingers (C++, Python and pybind11 modules)
- mpi_cmake_modules (containing CMake files with documentation)
- trifinger_simulation (pure Python package)

Future work I have in mind (but not in this PR):
- Option to auto-detect package version.
- Call the Python builder from CMake instead of duplicating the whole logic there.
- Option to use `doc/index.rst` as start page instead of the README

MaximilienNaveau commented on python/mpi_cmake_modules/documentation_builder.py:
```diff
@@ -480,48 +541,106 @@ def build_documentation(build_dir, project_source_dir, project_version):
+
+
+def main():
+    import argparse
```
should be in the header

MaximilienNaveau reviewed (changes requested):
One small change I asked. Otherwize good to go!
Thanks for the job.
I would love to extract this in another package than mpi_cmake_module :thinking: 
Something we can use independently maybe as git submodules could be nice... I still trying to wrap my head around the idea.

luator commented:
@MaximilienNaveau I agree, it would be nice to have this as an independent Python package.  A straight forward option would be to just specify it as a dependency of mpi_cmake_modules that needs to be installed first.
